### PR TITLE
tests/run-tests: Run tests inside a temporary chdir.

### PR DESCRIPTION
**Pull request by jimmo**
I think this is a good idea just for making sure the tests run in a self-contained environment, and possibly we should consider a unique temp dir per test invocation.

However, this is specifically to workaround a situation where I want to write a test for uasyncio (#6885), where uasyncio now does "import io", but this ends up importing tests/io instead.

I've tried to ensure that this doesn't break anything from #5877 (@stinos ), and in particular that running this from a out-of-tree directory continues to work.

### CODE REVIEW

**Comment by jimmo** [comment]
Updated to use a unique temp dir per invocaton. This is to allow copying required data files into the test dir. Any test in a directory that has a "data" directory will have this copied into the working dir.

File: tests/run-tests [hunk 1]
```diff
@@ -7,37 +7,58 @@ import platform
...
 # See stackoverflow.com/questions/2632199: __file__ nor sys.argv[0]
 # are guaranteed to always work, this one should though.
 BASEPATH = os.path.dirname(os.path.abspath(inspect.getsourcefile(lambda: None)))
 
+
+# Convert a path relative to micropython/tests to an absolute path.
 def base_path(*p):
-    return os.path.abspath(os.path.join(BASEPATH, *p)).replace('\\', '/')
+    return os.path.abspath(os.path.join(BASEPATH, *p)).replace("\\", "/")
+
```
Comment (stinos) [inline tests/run-tests:23]: Did you add these, or `black`? (just wondering)

Comment (jimmo) [reply tests/run-tests:23]: `black` -- The first commit is entirely black's changes (nothing from me).

**Comment by jimmo** [comment]
The underlying issue here was that run-tests.py used to do:

```py
os.environ["MICROPYPATH"] = os.pathsep + base_path("../extmod")
```

which was intended to make it possible to import frozen modules (and extmod modules), but had the additional undesirable effect of including TOP/tests in the module search path.

But now https://github.com/micropython/micropython/pull/8079 allows us to specifically add frozen modules to the path without using `""`, so this fix is no longer required.

**Pull request closed without merging**